### ISSUE TITLE
more unicode crashes :)

### ISSUE THREAD

**Issue by lowne** [Hammerspoon/hammerspoon#647]
``` lua
s='YÖLO'
d=drawing.text(geometry'0,0,100,100',s)
d:setText(s:sub(1,2))
```

(I guess I can use styledtext for rendering things properly, but here I just wish it didn't crash)

**Comment by lowne** [comment]
Also, speaking of styledtext:
```lua
> st=styledtext.new'YÖLO'

> inspect(st:sub(1,2):asTable())
{ "YÖ", {
    attributes = {},
    ends = 3, -- is it supposed to count bytes instead of codepoints?
    starts = 1
  } }

> st:sub(1,2):isIdentical(st:sub(1,3))
true -- huh

> st:byte()
...app/Contents/Resources/extensions/hs/styledtext/init.lua:170: attempt to call a nil value (method 'asString')

```

**Comment by asmagill** [comment]
the last is a typo missed after the name change for asString... the rest, I'll dig into tonight.

**Comment by asmagill** [comment]
Yes Lua works in terms of bytes, not codepoints, so s:sub(1,2) is an incomplete/invalid sequence as fas as UTF8 is concerned... this causes the crash in `d:setText(s:sub(1,2))`.  However, I don't know why `st:sub(1,2):isIdentical(st:sub(1,3))` is true... it shouldn't be.

Let me work on it... setting an invalid string shouldn't crash -- it should either error or silently truncate, not sure which will be easier yet, but it shouldn't crash.  And I'll look at the isIdentical method as well and see where it's failing.

**Comment by asmagill** [comment]
Ok... with pull #649, `hs.drawing:setText` and `hs.styledtext.new` don't crash when an invalid UTF8 sequence is encountered... instead they do what the console does and stick in the Unicode invalid character code.

There is a disconnect, however, in what `string:sub(i,j)` and `styledtext:sub(1,2)` do...

For strings, the `sub` method is in the Lua source.  It works based upon bytes, and can extract what ends up being invalid UTF8.  Doing this, and then assigning it to a drawing object as styled text will display the invalid UTF8 as the unicode invalid character, like the console, because the string which makes it into the Objective-C code is already broken (in terms of being valid UTF8).

For styledtext, the `sub` method is in Objective-C.  Objective-C doesn't do invalid UTF8 without first treating it as NSData and then doing a variety of checks before conversion... Objective-C either drops invalid UTF8 entirely (the entire string, not just the bad chars) or crashes.  Neither is acceptable.  Further, its methods all work in terms of codepoints, not in terms of bytes.  Because of the codepoint/byte thing, I use a mapping table so that the method arguments follow the Lua convention (bytes) and are converted to codepoints behind the scenes... because of this, if the full styledtext is a valid UTF8 sequence, but a substring, based on bytes, is not, it can't be properly represented... all portions of the byte sequence making up the UTF8 character convert to the same codepoint position in the styledtext.

Forcing this to act like Lua strings is not trivial... What this means is the following:

~~~lua
s='YÖLO'
st = hs.styledtext.new(s)

d = {
hs.drawing.rectangle{x=90,y=90,h=140,w=220}:setFillColor{}:show(),
hs.drawing.text({x=100,y=100,h=40,w=100},s):show(),
hs.drawing.text({x=100,y=140,h=40,w=100},""):setText(s:sub(1,2)):show(),
hs.drawing.text({x=100,y=180,h=40,w=100},""):setStyledText(s:sub(1,2)):show(),

hs.drawing.text({x=200,y=100,h=40,w=100},st):show(),
hs.drawing.text({x=200,y=180,h=40,w=100},""):setStyledText(st:sub(1,2)):show(),
}
~~~

Yields:
![screen shot 2015-11-20 at 12 22 36 pm](https://cloud.githubusercontent.com/assets/8139480/11308016/7bd1d81e-8f81-11e5-9d62-17c40ba9446e.png)

Is this a problem?

**Comment by lowne** [comment]
I haven't looked at the implementation; from what you write above, it seems that my assumptions were correct (styledtext built more on less on top of NSAttributedString, and NSAttributedString fully utf-8 aware, and therefore styledtext methods being codepoint-aware); I'm surprised to learn that you went out of your way to implement a byte-based mapping instead.
I understand the general reasoning behind keeping the same index mapping between Lua strings and NSStringWhatevers (wrt `:sub()` etc.), but off the top of my head I can't think of a compelling reason to do so in this specific case - why would I ever care about *bytes* in a styledtext? - while I can think of at least one for *not* doing it - namely, the fact that Lua strings being raw byte streams (or, equivalently, the lack of native and/or library provided, and included in HS, unicode character sequences) actually sucks when dealing with utf-8, which is *always* in UI-related matters.
Removing said mapping from `styledtext` would mean that HS (Lua) users finally gain (a bit out of left field) a utf-8-aware, integrated string library for "free" (even when one doesn't care about styling).


TL;DR: I don't care about styledtext objects acting like Lua strings, because Lua strings can't be (natively) used to manipulate utf-8 sanely anyway. I'd rather they do the expected thing, which is for `'YÖLO':sub(1,3)` to be `YÖL`. But that's just my 2c.

**Comment by asmagill** [comment]
The reason for doing it the way I did was because, ultimately, the environment the user is using is Lua, not Objective-C.  The ansi converter is written in Lua... it calculates where to put colors based upon code sequences... identified in Lua... if I hadn't done the mapping internally, then the converter (or any other) would have had to... it's why the first version of the Ansi conversion tools failed when UTF8 was introduced.

However, you do make some interesting points, if we want to extend it a bit and make it a true "UTF8 aware string library".

**Comment by lowne** [comment]
Ah, yes, I didn't account for the ansi escapes, which are the one case where the actual bytes matter :cold_sweat:

**Comment by cmsj** [comment]
I'm going to close this because I'm gardening the bugs and this has been open for ages with no further discussion, feel free to re-open it if there are specific things you want to do :)
